src/form/remaining-chars-count/remaining-chars-count.directive.js

```diff
@@ -0,0 +1,177 @@
+     $scope.messageArea2text = "";
+     $scope.messageInput3text = "";
+
+     $scope.charsMaxLimitExceeded = false;
+
+     // 'tfId' will equal the id of the text area/input field which
+     // triggered the event
+     $scope.$on('overCharsMaxLimit', function (event, tfId) {
+         if(!$scope.charsMaxLimitExceeded){
+           $scope.charsMaxLimitExceeded = true;
+           $scope.$digest();
+         }
+     });
+
+     // 'tfId' will equal the id of the text area/input field which
+     // triggered the event
+     $scope.$on('underCharsMaxLimit', function (event, tfId) {
+         if($scope.charsMaxLimitExceeded){
+           $scope.charsMaxLimitExceeded = false;
+           $scope.$digest();
+         }
+     });
+
+   });
+
+   </file>
+ </example>
+*/
+
+angular.module('patternfly.form').directive('pfRemainingCharsCount', function ($timeout) {
+  'use strict';
+  return {
+    restrict: 'A',
+    require: 'ngModel',
+    scope: {
+      ngModel: "="
+    },
+    link: function ($scope, $element, $attributes) {
+      var charsMaxLimit = $attributes.charsMaxLimit;
+      var charsWarnRemaining = $attributes.charsWarnRemaining;
+      var countRemainingFld = angular.element(document.getElementById($attributes.countFld));
+      var blockInputAtMaxLimit = ($attributes.blockInputAtMaxLimit === 'true');
+      var charsRemainingFn = function () {
+        var charsLength = $scope.ngModel.length;
+        var remainingChars = charsMaxLimit - charsLength;
+
+        // trim if blockInputAtMaxLimit and over limit
+        if (blockInputAtMaxLimit && charsLength > charsMaxLimit) {
+          $scope.ngModel = $scope.ngModel.substring(0, charsMaxLimit);
+          charsLength = $scope.ngModel.length;
+          remainingChars = charsMaxLimit - charsLength;
+        }
+
+        // creating scope vars for unit testing
+        $scope.remainingChars = remainingChars;
+        $scope.remainingCharsWarning = (remainingChars <= charsWarnRemaining ? true : false);
+
+        countRemainingFld.text(remainingChars);
+        countRemainingFld.toggleClass('chars-warn-remaining-pf', remainingChars <= charsWarnRemaining);
+
+        // dispatching the $emit immediately throws a 'already in $digest' error, so waiting a half-sec
+        $timeout(function () {
+          if (remainingChars < 0) {
+            $scope.$emit('overCharsMaxLimit', $attributes.id);
+          } else {
+            $scope.$emit('underCharsMaxLimit', $attributes.id);
+          }
+        }, 100);
+      };
+
+      $scope.$watch('ngModel', function () {
+        charsRemainingFn();
+      });
+
+      $element.bind('keypress', function (event) {
+        // Once the charsMaxLimit has been met or exceeded, prevent all keypresses from working
+        if (blockInputAtMaxLimit && $element.val().length >= charsMaxLimit) {
+          // Except backspace
+          if (event.keyCode !== 8) {
+            event.preventDefault();
+          }
```

Comment:
This does prevent the Delete key, but backspace delete is available.
This does handle cut & paste, both from keys and right-click.